### PR TITLE
feat: add support for fixed list wildcard in type signature

### PR DESCRIPTION
**Pull request by universalmind303**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/9139

## Rationale for this change
better support for working with fixed size lists is needed

### CODE REVIEW

**Comment by jayzhan211** [comment]
Shouldn't fixed size list be converted to list? @universalmind303 Is there any example / test that shows the need.

**Comment by jayzhan211** [comment]
arrow cast does not support to cast to wildcard size, after coerce_from, how do you process the wildcard size?

**Comment by universalmind303** [comment]
> arrow cast does not support to casting to wildcard size, after `coerce_from` the original size info is lost, how do you process the wildcard size instead

I think i actually had this switched around. 

`FixedSizeList(_, FIXED_SIZE_LIST_WILDCARD))` should coerce to `FixedSizeList(_, size))` not the other way around.

File: datafusion/expr/src/type_coercion/functions.rs [hunk 1]
```diff
@@ -372,13 +374,20 @@ fn coerced_from<'a>(
...
         List(_) | LargeList(_)
             if datafusion_common::utils::base_type(type_from).eq(&Null)
                 || list_ndims(type_from) == list_ndims(type_into) =>
         {
             Some(type_into.clone())
         }
+        // should be able to coerce wildcard fixed size list to non wildcard fixed size list
+        FixedSizeList(_, _)
+            if matches!(type_from, FixedSizeList(_, FIXED_SIZE_LIST_WILDCARD))
+                && list_ndims(type_from) == list_ndims(type_into) =>
```
Comment (gruuya) [inline datafusion/expr/src/type_coercion/functions.rs:387]: Perhaps there should also be a condition for checking coerce-ability of the base types here (and maybe a negative test too), something along the lines of:
 ```rust
... && coerced_from(base_type(type_into), base_type(type_from)) == Some(base_type(type_into))
```

**Comment by jayzhan211** [comment]
If I understand correctly, it should pass the test like

```rust
        let inner = Arc::new(Field::new("item", DataType::Int32, false));
        let current_types = vec![
            DataType::FixedSizeList(inner.clone(), 2), // able to coerce for any size
        ];
        let signature = Signature::exact(vec![
            DataType::FixedSizeList(inner.clone(), FIXED_SIZE_LIST_WILDCARD),
        ], Volatility::Stable);
        let coerced_data_types = data_types(&current_types, &signature).unwrap();
        assert_eq!(coerced_data_types, current_types);
```

File: datafusion/expr/src/type_coercion/functions.rs (hunk 1 above)
Comment (gruuya) [reply datafusion/expr/src/type_coercion/functions.rs:387]: Ok got a bit nerd-sniped here it seems, but I think the following actually handles the nested `FixedSizeLists` too:
```rust
FixedSizeList(f_into, FIXED_SIZE_LIST_WILDCARD) => match type_from {
    FixedSizeList(f_from, size_from) => {
        match coerced_from(f_into.data_type(), f_from.data_type()) {
            Some(data_type) if &data_type != f_into.data_type() => {
                let new_field =
                    Arc::new(f_into.as_ref().clone().with_data_type(data_type));
                Some(FixedSizeList(new_field, *size_from))
            }
            Some(_) => Some(FixedSizeList(f_into.clone(), *size_from)),
            _ => None,
        }
    }
    _ => None,
},
```

The following test passes (as does the one by @jayzhan211 ):
```rust
#[test]
fn test_nested_wildcard_fixed_size_lists() -> Result<()> {
    let type_into = DataType::FixedSizeList(
        Arc::new(Field::new(
            "item",
            DataType::FixedSizeList(
                Arc::new(Field::new("item", DataType::Int32, false)),
                FIXED_SIZE_LIST_WILDCARD,
            ),
            false,
        )),
        FIXED_SIZE_LIST_WILDCARD,
    );

    let type_from = DataType::FixedSizeList(
        Arc::new(Field::new(
            "item",
            DataType::FixedSizeList(
                Arc::new(Field::new("item", DataType::Int8, false)),
                4,
            ),
            false,
        )),
        3,
    );

    assert_eq!(
        coerced_from(&type_into, &type_from),
        Some(DataType::FixedSizeList(
            Arc::new(Field::new(
                "item",
                DataType::FixedSizeList(
                    Arc::new(Field::new("item", DataType::Int32, false)),
                    4,
                ),
                false,
            )),
            3,
        ))
    );

    Ok(())
}
```

**Comment by universalmind303** [comment]
@jayzhan211 it should be ready for review now.

also @gruuya thanks for the suggestions. That snippet seems to get all tests to pass!

**Comment by universalmind303** [comment]
test failures seem unrelated. I get same errors on main

**Comment by gruuya** [comment]
One last thing that might be missing is handling the non-wildcard scenario, i.e. when someone might want to specify an specific/explicit size for the `FixedSizeList` in the signature.

**Comment by universalmind303** [comment]
> One last thing that might be missing is handling the non-wildcard scenario, i.e. when someone might want to specify an specific/explicit size for the `FixedSizeList` in the signature.

I just added a test case for that.

File: datafusion/expr/src/type_coercion/functions.rs [hunk 2]
```diff
@@ -492,4 +515,77 @@ mod tests {
...
+        let coerced_data_types = data_types(&current_types, &signature).unwrap();
+        assert_eq!(coerced_data_types, current_types);
+
+        // make sure it can't coerce to a different size
+        let signature = Signature::exact(
+            vec![DataType::FixedSizeList(inner.clone(), 3)],
+            Volatility::Stable,
+        );
+        let coerced_data_types = data_types(&current_types, &signature);
+        assert!(coerced_data_types.is_err());
```
Comment (gruuya) [inline datafusion/expr/src/type_coercion/functions.rs:543]: This is a good assert, though what i had in mind is that it probably wouldn't work for `current_types = vec![DataType::FixedSizeList(inner.clone(), 3)]`, even though in that case it should, right?

Comment (universalmind303) [reply datafusion/expr/src/type_coercion/functions.rs:543]: so like making sure the same type works? 

```rs
        let current_types = vec![
            DataType::FixedSizeList(inner.clone(), 2), // able to coerce for any size
        ];

        // make sure it works with the same type.
        let signature = Signature::exact(
            vec![DataType::FixedSizeList(inner.clone(), 2)],
            Volatility::Stable,
        );
        let coerced_data_types = data_types(&current_types, &signature).unwrap();
        assert_eq!(coerced_data_types, current_types);
```

Comment (gruuya) [reply datafusion/expr/src/type_coercion/functions.rs:543]: Yes, exactly.

**Comment by alamb** [comment]
@gruuya  do you think this PR is good to go?

**Comment by gruuya** [comment]
Yup, looks good!

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @universalmind303  for the contribution and @gruuya  for the review

One thing that might be valuable to do is to write some sort of test (perhaps a UDF in https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/user_defined/user_defined_scalar_functions.rs) that used this coercion rule and showed it being called with `FixedSizedLists` of difference sizes. 

I was thinking that way the "end to end" desired behavior would also be encoded in a test.

However, I think this PR already has adequate tests, so we can add additional tests as a follow on PR if we want

**Pull request merged**